src/Cpu.re

```diff
@@ -190,20 +323,30 @@ let step_size = (definition: Instruction.t, opcode: Opcode.t) => {
   };
 };
 
+let rmw_update = (opcode: Opcode.t, address) => {
+  switch (opcode.addressing_mode) {
+  | AddressingMode.Accumulator => (cpu: t, value) => cpu.acc = value
+  | _ => (cpu: t, value) => Memory.set_byte(cpu.memory, address, value)
+  };
+};
+
 let handle = (definition: Instruction.t, opcode: Opcode.t, cpu: t) => {
   let address = AddressingMode.get_address(cpu, opcode.addressing_mode);
 
   let operand =
-    switch (definition.access_pattern) {
-    | Instruction.Read => Memory.get_byte(cpu.memory, address)
-    | Instruction.Static => 0
-    | _ => address
+    switch (definition.access_pattern, opcode.addressing_mode) {
+    | (Instruction.Static, _) => 0
+    | (Instruction.Write, _) => address
+    | (Instruction.Jump, _) => address
+    | (Instruction.ReadModifyWrite, AddressingMode.Accumulator) => address
+    | _ => Memory.get_byte(cpu.memory, address)
     };
 
   let operation =
     switch (definition.label) {
     | "adc" => add_with_carry
     | "and" => and_with_acc
+    | "asl" => shift_left(rmw_update(opcode, address))
```

Comment:
For what it's worth, I was getting a similar sense writing the several `verb_(acc|x|y)` instructions. What I _want_ here is to have a `x, y, acc: lens(cpu, int) = { get: (cpu) => int, set: (cpu, int) => cpu }` for each of those and write something like:

```
| "cmp" => comapre(acc)
| "cpx" => compare(x)
| "cpy" => compare(y)
```

Maybe worth looking into something like [this](https://github.com/jonlaing/rationale/blob/master/src/Lens.re)? Maybe there's a `opcode => lens(cpu, int)` somewhere?

Reply:
Love this idea. I'm not familiar with Lenses and of a mind to put off refactorings until we've cranked through all the legal opcode tests but in a day or two I'd love to learn how lenses do. :)